Update library to use JS Promises instead of callbacks

Steague opened the pull request:
Now methods can be chained together.

This PR contains breaking changes. All `callback` logic has been removed and replaced with Promise logic.

Unit tests have been updated.

Steague commented:
@todbot 

Node.js added native promise support since version 0.11.13. (Which was released in May, 2014.)
Node.js added native ES6 support since version 4.3.2. (Which was released in March, 2016.)

I would consider neither 0.11 nor 4.3 to be slightly older and are, in fact, significantly old.

Since the parameters of the functions have been updated to accept an object instead of many parameters, adding a callback property to that object would not be difficult but I see it as redundant ad backwards-compatibility is already not an option with this PR.

sandeepmistry commented:
@Steague thanks for submitting!

I agree with @todbot, I'd prefer not to have breaking changes and force users into using promises.

Some proposals to go forward:

1) Have a backwards compatible callback mode as discussed above

2) Have something like `var Blink1 = require('node-blink1).Promise` where users can opt-into promises

3) Have another npm module for Blink1 with promises.

Steague commented:
@sandeepmistry & @todbot I understand where you guys are coming from. I'll make my own API in these cases.

By the way, Node 4.x is only going to be supported for the next 3.5 months. https://github.com/nodejs/Release

I would recommend determining backwards compatibility with the current code before moving to a 1.0 release.

sandeepmistry commented:
@todbot the proposal sounds good to me, pull requests welcome for it :)

Another option could be to always make the promise, and add a `.then(callback)` if user provides a callback.

Steague commented:
@todbot && @sandeepmistry I have added callbacks back to the code base. They are called automatically just before the promises are returned from the various class methods. The README has been updated to reflect these changes.

Additionally, I split the code up a bit to separate the validators, helpers, and base hardware API from the main methods being used.

I also added the rgbcolor package to the codebase as a supported color parameter. It is explained in the README and is not required. This allows users to specify colors like "red" instead of needing to know the RGB values.

sandeepmistry commented:
So this will still have breaking changes to the API right? Some parameters are an object now instead of multiple parameters.

@todbot opinions on this? It's going to be a breaking change, so will require a major rev update for semver, if Node 6 is the minimum version as well.

fushi commented:
Is this still being considered? Promises are a core feature of the language now.
